roscala/src/main/scala/coop/rchain/rosette/Ctxt.scala

```diff
@@ -1,42 +1,102 @@
 package coop.rchain.rosette
 
+import coop.rchain.rosette.Location.StoreCtxt
+import Location._
 import scala.collection.mutable
 
-case class Ctxt(argvec: Tuple,
-                code: Code,
-                ctxt: Ctxt,
-                env: Ob,
-                monitor: Monitor,
+case class Ctxt(tag: Location,
                 nargs: Int,
                 outstanding: Int,
                 pc: PC,
-                reg: Seq[Ob],
-                rslt: Ob,
-                trgt: Ob,
-                selfEnv: Ob,
-                override val _slot: mutable.Seq[Ob],
-                tag: Location)
+                rslt: Ob, // reg[0]
+                trgt: Ob, // reg[1]
+                argvec: Tuple, // reg[2]
+                env: Ob, // reg[3]
+                code: Code, // reg[4]
+                ctxt: Ctxt, // reg[5]
+                self2: Ob, // reg[6]
+                selfEnv: Ob, // reg[7]
+                rcvr: Ob, // reg[8]
+                monitor: Monitor, // reg[9]
+                override val _slot: mutable.Seq[Ob])
     extends Ob {
+  def applyK(result: Ob, tag: Location)(state: VMState): (Boolean, VMState) =
+    ctxt.rcv(result, tag)(state)
+
   def arg(n: Int): Option[Ob] = argvec.elem.lift(n)
 
+  def getReg(r: Int): Option[Ob] =
+    r match {
+      case 0 => Some(rslt)
+      case 1 => Some(trgt)
+      case 2 => Some(argvec)
+      case 3 => Some(env)
+      case 4 => Some(code)
+      case 5 => Some(ctxt)
+      case 6 => Some(self2)
+      case 7 => Some(selfEnv)
+      case 8 => Some(rcvr)
+      case 9 => Some(monitor)
+      case _ => None
+    }
+
   /** This is necessary because the compiler sometimes arranges to
     *  provide an argvec that is acually longer than nargs indicates. If
     *  we are about to expose the context to the outside world, we need
     *  to clean it up so that it appears consistent.
     */
   def prepare(): Ctxt = this.copy(argvec = argvec.makeSlice(0, nargs))
 
-  def ret(rslt: Ob): Boolean = true
+  def rcv(result: Ob, loc: Location)(state: VMState): (Boolean, VMState) = {
+    val storeResult = Location.store(loc, this, state.globalEnv, result)
+
+    storeResult match {
+      case StoreCtxt(newCtxt) =>
```

Comment:
Hmm?

Reply:
Around the case block when it's not a single expression.

Reply:
Is that in our style guide? If so, I would vote for removing that.
See: http://docs.scala-lang.org/style/control-structures.html#curly-braces

Reply:
It's not in the style guide per se, but we do use braces even in situations where they're not strictly required:
```scala
if (cond) {
  result
} else {
  otherResult
} 
for (x <- y if cond) yield {
  operation(x)
}

while (notDone()) {
  doTheThing()
}
```

Reply:
I guess according to the Scala style guide we should also do:
```
if (cond)
  result
else
  otherResult
```
Personally I don't mind if we don't.

I still think we should omit braces for case clauses. Even IntelliJ will give you a hint to remove redundant braces.
Do you insist on the braces?

Reply:
I suppose case is different than if, since adding more expressions below it will be considered part of the same case, while adding more expressions below an if won't be:

```scala
case foo(x) =>
    y = x+1
    y*y
```
vs
```scala
if (cond)
    y = x+1
    y*y
```